README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

[![Lifecycle:
stable](https://img.shields.io/badge/lifecycle-stable-brightgreen.svg)](https://lifecycle.r-lib.org/articles/stages.html#stable)
[![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/nhdR)](https://cran.r-project.org/package=nhdR)
[![R-CMD-check](https://github.com/jsta/nhdR/actions/workflows/R-CMD-check.yml/badge.svg)](https://github.com/jsta/nhdR/actions/workflows/R-CMD-check.yml)
[![DOI](https://zenodo.org/badge/75339263.svg)](https://zenodo.org/badge/latestdoi/75339263)

# nhdR <a href='https://jsta.github.io/nhdR/'><img src="man/figures/logo.png" align="right" height=140/></a>

Tools for querying, downloading, and networking both the [National
Hydrography Dataset (NHD)](https://www.usgs.gov/national-hydrography)
and
[NHDPlus](https://www.epa.gov/waterdata/nhdplus-national-hydrography-dataset-plus)
datasets.

## Installation

CRAN policy is that no package can write to a persistent location by
default. As a result, `nhdR` writes all data to a temporary location
unless a `temporary = FALSE` argument is passed to the
`nhd_plus_get`/`nhd_get` functions. Alternatively, `nhdR` will
automatically write data to a persistent location if the `nhdR_path`
environment variable is set. To do this, add the following line to your
`.Rprofile`:

``` r
Sys.setenv(nhdR_path = file.path(rappdirs::user_data_dir(appname = "nhdR",
      appauthor = "nhdR")))
```

Your `.Rprofile` file can be edited using the
`usethis::edit_r_profile()` function.

### Stable version from CRAN

``` r
install.packages("nhdR")
```

### or development version from GitHub

``` r
# install.packages("devtools")
devtools::install_github("jsta/nhdR")
```

This package also requires an installation of
[7-zip](https://www.7-zip.org/) that can be called via the command line
using `7z` or `7za.exe` (check if your machine is good to go with
`nhdR:::has_7z()`).

## Usage

### Load package

``` r
library(nhdR)
```

### NHD Plus

NHD-Plus exports are organized by vector processing unit (vpu). See
below for a low resolution vpu map (also `nhdR::vpu_shp`). A hi-res
version can be found
[here](https://www.epa.gov/waterdata/nhdplus-global-data).

![](man/figures/README-unnamed-chunk-4-1.png)<!-- -->

``` r
# get a vpu export
nhd_plus_get(vpu = 4, "NHDSnapshot")
nhd_plus_get(vpu = 4, "NHDPlusAttributes")
nhd_plus_get(vpu = 4, "NHDPlusCatchment")
```

``` r
# list layers
nhd_plus_list(vpu = 4, "NHDSnapshot")
#>  [1] "NHDArea.dbf"                "NHDAreaEventFC.dbf"        
#>  [3] "NHDAreaEventFC.shp"         "NHDArea.shp"               
#>  [5] "NHDFCode.dbf"               "NHDFlowline.dbf"           
#>  [7] "NHDFlowline.shp"            "NHDFlowline.shp.xml"       
#>  [9] "NHDLine.dbf"                "NHDLineEventFC.dbf"        
#> [11] "NHDLineEventFC.shp"         "NHDLine.shp"               
#> [13] "NHDPoint.dbf"               "NHDPointEventFC.dbf"       
#> [15] "NHDPointEventFC.shp"        "NHDPoint.shp"              
#> [17] "NHDReachCode_Comid.dbf"     "NHDReachCrossReference.dbf"
#> [19] "NHDWaterbody.dbf"           "NHDWaterbody.shp"
nhd_plus_list(vpu = 4, "NHDPlusAttributes")
#>  [1] "CumulativeArea.dbf"              "DivFracMP.dbf"                  
#>  [3] "elevslope.dbf"                   "HeadwaterNodeArea.dbf"          
#>  [5] "MegaDiv.dbf"                     "PlusARPointEvent.dbf"           
#>  [7] "PlusFlowAR.dbf"                  "PlusFlow.dbf"                   
#>  [9] "PlusFlowlineLakeMorphology.dbf"  "PlusFlowlineVAA.dbf"            
#> [11] "PlusWaterbodyLakeMorphology.dbf"
nhd_plus_list(vpu = 4, "NHDPlusCatchment")
#> [1] "Catchment.dbf"         "Catchment.shp"         "featureidgridcode.dbf"
```

``` r
# get layer info
nhd_plus_info(vpu = 4, "NHDSnapshot", "NHDWaterbody")
```

    #>  [1] "Driver: ESRI Shapefile; number of rows: 31830 "     
    #>  [2] "Feature type: wkbPolygon with 3 dimensions"         
    #>  [3] "Extent: (-93.24332 40.43575) - (-73.61814 48.11344)"
    #>  [4] "CRS: +proj=longlat +datum=NAD83 +no_defs  "         
    #>  [5] "LDID: 87 "                                          
    #>  [6] "Number of fields: 12 "                              
    #>  [7] "         name type length typeName"                 
    #>  [8] "1       COMID    0      9  Integer"                 
    #>  [9] "2       FDATE    9     10     Date"                 
    #> [10] "3  RESOLUTION    4      7   String"                 
    #> [11] "4     GNIS_ID    4     10   String"                 
    #> [12] "5   GNIS_NAME    4     65   String"                 
    #> [13] "6    AREASQKM    2     19     Real"                 
    #> [14] "7   ELEVATION    2     19     Real"                 
    #> [15] "8   REACHCODE    4     14   String"                 
    #> [16] "9       FTYPE    4     24   String"                 
    #> [17] "10      FCODE    0      9  Integer"                 
    #> [18] "11 SHAPE_LENG    2     19     Real"                 
    #> [19] "12 SHAPE_AREA    2     19     Real"

``` r
# load layer
dt <- nhd_plus_load(vpu = 4, "NHDSnapshot", "NHDWaterbody")
#> Reading layer `NHDWaterbody' from data source 
#>   `/home/jemma/.local/share/nhdR/NHDPlus/GL_04_NHDSnapshot/NHDWaterbody.shp' 
#>   using driver `ESRI Shapefile'
#> Simple feature collection with 31830 features and 12 fields
#> Geometry type: POLYGON
#> Dimension:     XYZ
#> Bounding box:  xmin: -93.24332 ymin: 40.43575 xmax: -73.61814 ymax: 48.11344
#> z_range:       zmin: 0 zmax: 0
#> Geodetic CRS:  NAD83
```

### NHD

NHD exports are organized by US state.

``` r
nhd_get(state = c("DC", "HI"))
```

``` r
nhd_list(state = "DC")
#>  [1] "ExternalCrosswalk"           "NHDFCode"                   
#>  [3] "NHDFeatureToMetadata"        "NHDFlow"                    
#>  [5] "NHDFlowlineVAA"              "NHDMetadata"                
#>  [7] "NHDProcessingParameters"     "NHDReachCodeMaintenance"    
#>  [9] "NHDReachCrossReference"      "NHDSourceCitation"          
#> [11] "NHDStatus"                   "NHDVerticalRelationship"    
#> [13] "NHDPoint"                    "NHDFlowline"                
#> [15] "NHDLine"                     "NHDArea"                    
#> [17] "NHDWaterbody"                "NHDAreaEventFC"             
#> [19] "NHDLineEventFC"              "NHDPointEventFC"            
#> [21] "WBDLine"                     "NonContributingDrainageArea"
#> [23] "NWISBoundary"                "NWISDrainageArea"           
#> [25] "WBDHU14"                     "WBDHU8"                     
#> [27] "WBDHU2"                      "WBDHU4"                     
#> [29] "WBDHU6"                      "WBDHU10"                    
#> [31] "WBDHU12"                     "WBDHU16"                    
#> [33] "HYDRO_NET_Junctions"        
#> attr(,"driver")
#> [1] "OpenFileGDB"
#> attr(,"nlayers")
#> [1] 33
```

``` r
nhd_info(state = "DC", dsn = "NHDWaterbody")
#> Source: "/home/jemma/.local/share/nhdR/NHD_H_District_of_Columbia_State_GDB.gdb", layer: "NHDWaterbody"
#> Driver: OpenFileGDB; number of rows: 8011 
#> Feature type: wkbPolygon with 3 dimensions
#> Extent: (-78.07095 38.52142) - (-76.82219 39.64683)
#> CRS: +proj=longlat +datum=NAD83 +no_defs 
#> Number of fields: 13 
#>                    name type length typeName
#> 1  Permanent_Identifier    4     40   String
#> 2                 FDate   11      0 DateTime
#> 3            Resolution    0      0  Integer
#> 4               GNIS_ID    4     10   String
#> 5             GNIS_Name    4     65   String
#> 6              AreaSqKm    2      0     Real
#> 7             Elevation    2      0     Real
#> 8             ReachCode    4     14   String
#> 9                 FType    0      0  Integer
#> 10                FCode    0      0  Integer
#> 11     VisibilityFilter    0      0  Integer
#> 12         Shape_Length    2      0     Real
#> 13           Shape_Area    2      0     Real
```

``` r
head(nhd_load(state = "DC", dsn = "NHDWaterbody"))
#> Reading layer `NHDWaterbody' from data source 
#>   `/home/jemma/.local/share/nhdR/NHD_H_District_of_Columbia_State_GDB.gdb' 
#>   using driver `OpenFileGDB'
#> Simple feature collection with 8011 features and 13 fields
#> Geometry type: MULTIPOLYGON
#> Dimension:     XYZ
#> Bounding box:  xmin: -78.07095 ymin: 38.52142 xmax: -76.82219 ymax: 39.64683
#> z_range:       zmin: 0 zmax: 0
#> Geodetic CRS:  NAD83
#> Reading query `SELECT * from NHDWaterbody LIMIT 1' from data source `/home/jemma/.local/share/nhdR/NHD_H_District_of_Columbia_State_GDB.gdb' 
#>   using driver `OpenFileGDB'
#> Simple feature collection with 1 feature and 13 fields
#> Geometry type: MULTIPOLYGON
#> Dimension:     XYZ
#> Bounding box:  xmin: -76.99652 ymin: 38.68957 xmax: -76.99631 ymax: 38.6897
#> z_range:       zmin: 0 zmax: 0
#> Geodetic CRS:  NAD83
#> Simple feature collection with 6 features and 13 fields
#> Geometry type: POLYGON
#> Dimension:     XY
#> Bounding box:  xmin: -77.5767 ymin: 38.68957 xmax: -76.99631 ymax: 39.5882
#> Geodetic CRS:  WGS 84
#>   Permanent_Identifier               FDate Resolution GNIS_ID GNIS_Name
#> 1             46565431 2002-07-21 18:00:00          2    <NA>      <NA>
#> 2             51767181 2002-08-14 18:00:00          2    <NA>      <NA>
#> 3             51767223 2002-08-14 18:00:00          2    <NA>      <NA>
#> 4             51767287 2002-08-14 18:00:00          2    <NA>      <NA>
#> 5             51767709 2002-08-14 18:00:00          2    <NA>      <NA>
#> 6             51768273 2002-08-14 18:00:00          2    <NA>      <NA>
#>   AreaSqKm Elevation      ReachCode FType FCode VisibilityFilter Shape_Length
#> 1    0.000        NA 02070010004605   436 43624                0 0.0005402029
#> 2    0.002        NA 02070008004808   390 39004            50000 0.0017289109
#> 3    0.001        NA 02070008004829   390 39004          2000000 0.0013369633
#> 4    0.001        NA 02070008004860   390 39004            24000 0.0011083831
#> 5    0.002        NA 02070008005063   390 39004            50000 0.0016429957
#> 6    0.001        NA 02070008005335   390 39004            24000 0.0012442057
#>     Shape_Area                          Shape
#> 1 1.879174e-08 POLYGON ((-76.99631 38.6896...
#> 2 1.954519e-07 POLYGON ((-77.56946 39.5881...
#> 3 1.239613e-07 POLYGON ((-77.56954 39.5567...
#> 4 8.130533e-08 POLYGON ((-77.57658 39.5250...
#> 5 1.745505e-07 POLYGON ((-77.46919 39.3298...
#> 6 8.126193e-08 POLYGON ((-77.2087 39.18799...
```


NEWS.md
## nhdR 0.6.1

* Update nhdplus download endpoint (#92)

## nhdR 0.6.0

* Dropped use of the retiring `rgdal` package (#91)

* Added NHD support for AK, missing territories (#90), thanks @hrecht !

## nhdR 0.5.9

* Fix upstream dependency issue with the `digest` package (#89)

## nhdR 0.5.8

* Updated nhd and nhdplus downloading endpoints

## nhdR 0.5.7

* `nhd_query` now accepts polygon queries (#85)
* Enabled speedier `nhd_load`(ing) with `wkt_filter` and `query` arguments (#85)

## nhdR 0.5.6

* CRAN maintenance release
* Added support for recent `sf` package versions depending on `s2` geometry

## nhdR 0.5.5

* Drop automatic conversion to Geopackage for NHD (non-plus) products
* Massive (~30x) speed improvement for `nhd_plus` query operations (#63)
* Bug fix for loading `NHDReachCrossReference` layers (#72)
* Bug fix to make `vpu_shp` geometries "valid" (#70)

## nhdR 0.5.4

* Update URL endpoint for NHDPlus (#64)
* Bug fix for `nhd_plus_get` unzipping to enable the `force_unzip` argument

## nhdR 0.5.3

* Update URL endpoint for NHDPlus
* Fix crs tests for upcoming sf release

## nhdR 0.5.2

* Bug fix for multi-vpu data loading with mismatched columns (#57)
* Bug fix for character VPUs that match multiple remote data sources (#58)
* Added package logo

## nhdR 0.5.1

* Critical bug fix for `nhd_get` recursive folder creation (# 56)
* Bug fixes for `extract_network` enabling propagation of `approve_all_dl` flag
* Improved documentation of non-geographic buffer extraction

## nhdR 0.5.0

* Improve `extract_network` to handle headwater lakes
* Add function to return Great Lakes polygons
* Enable non-geographic query buffers
* Fix broken NHD download endpoint
* Bug fixes for `nhd_plus_query` on multi vpu queries

## nhdR 0.4

* Confirmed working operation on MacOS
* Results of load functions are now cached (memoized) for massive increase in package speed

## nhdR 0.3

* Added a `NEWS.md` file to track changes to the package.


R/utils.R
nhd_path <- function(temporary = TRUE) {
  if (nchar(Sys.getenv("nhdR_path")) != 0) {
    path <- Sys.getenv("nhdR_path")
    return(path)
  }

  if (temporary) {
    warning(
      "Recommended: set the 'temporary' argument to FALSE to save data to a
      persistent rappdirs location.")
    path <- tempdir()
    Sys.setenv(nhdR_path = path)
    return(path)
  } else {
    path <- file.path(rappdirs::user_data_dir(appname = "nhdR",
      appauthor = "nhdR"))
    dir.create(path, showWarnings = FALSE, recursive = TRUE)
    Sys.setenv(nhdR_path = path)
    return(path)
  }
}

gdb_path <- function(state) {
  paste0(nhd_path(), .Platform$file.sep, "NHD_H_", stateabb2name(state), "_State_GDB.gdb")
}

gdb_plus_path <- function() {
  file.path(nhd_path(),
    "NHDPlusV21_NationalData_National_Seamless_Geodatabase_02",
    "NHDPlusNationalData",
    "NHDPlusV21_National_Seamless.gdb")
}

# zip_plus_path <- function(){
#   file.path
#
# }

get_if_not_exists <- function(url, destfile, force_dl = FALSE) {
  if (force_dl) {
    message(paste0("Re-downloading ", url))
    suppressWarnings(
      httr::GET(url, httr::write_disk(destfile, overwrite = TRUE), httr::progress()))
    TRUE
  }

  if (!file.exists(destfile)) {
    message(paste0("Downloading ", url))
    suppressWarnings(
      httr::GET(url, httr::write_disk(destfile), httr::progress()))
    TRUE
  } else {
    message(paste0("A local copy of ", url, " already exists on disk"))
    FALSE
  }
}

zero_pad <- function(x, digits) {
  if (nchar(stringr::str_extract(x, "\\d+")) < 2) {
    paste0(
      paste0(
        rep(0, digits), collapse = ""), x, collapse = "")
  } else {
    x
  }
}

get_remotepath <- function(state, baseurl) {
  # state  <- "Missouri"
  # baseurl <- paste0("https://prd-tnm.s3.amazonaws.com/StagedProducts/Hydrography/NHD/State/")
  filename <- paste0("NHD_H_", state, "_State_GDB.zip")
  url      <- paste0(baseurl, "GDB/", filename)
  list(filename = filename, url = url)
}

# get_plus_remotepath("9")
#' @importFrom xml2 read_xml
get_plus_remotepath <- function(vpu, component = "NHDSnapshot") {

  if (vpu == "National") {
    prefix <- "NHDPlusV21/Data/NationalData/"
  } else {
    vpu_key <- readRDS(system.file("vpu_key.rds", package = "nhdR"))
    prefix <- vpu_key[vpu_key$vpu == zero_pad(vpu, 1), "directory"]
  }

  baseurl <- paste0(
    "https://dmap-data-commons-ow.s3.amazonaws.com?delimiter=/&prefix=",
    prefix)
  res <- read_xml(baseurl)
  res <- stringr::str_extract_all(as.character(res),
    "(?<=\\<Key>).*(?=<\\/Key>)")[[1]]

  if (vpu == "National") {
    res <- unlist(res[grep(component, res)])
  } else {
    res <- unlist(res[grep(paste0(vpu, "_", component), res)])
  }

  res <- res[!(seq_len(length(res)) %in%
    c(grep("FGDB", res), grep(".pdf", res), grep("FileGDB", res)))]

  paste0("https://dmap-data-commons-ow.s3.amazonaws.com/", res[1])
}

is_spatial <- function(filename) {
  length(grep("shp$", filename)) > 0
}

#' Find VPU
#'
#' Find Vector Processing Unit from sf object
#'
#' @param pnt sf object
#'
#' @importFrom sf st_transform st_crs st_join st_distance
#' @importFrom dplyr select
#' @export
#' @return A character vector of vpu ids
#'
#' @examples \dontrun{
#' library(sf)
#'
#' # vpu centers
#' pnt <- st_cast(st_point_on_surface(nhdR::vpu_shp), "POINT")
#'
#' find_vpu(pnt[1, ])
#' find_vpu(pnt)
#'
#' find_vpu(nhdR::gull$sp$NHDWaterbody[1, ])
#' find_vpu(nhdR::gull$sp$NHDWaterbody)
#' }
find_vpu <- function(pnt) {
  # fix for older proj versions (solaris)
  # https://stackoverflow.com/a/62268361/3362993
  vpu <- nhdR::vpu_shp
  sf::st_crs(vpu$geometry) <- 4326

  pnt <- sf::st_transform(pnt, sf::st_crs(vpu))
  vpu <- vpu[vpu$UnitType == "VPU", ]

  if (any(names(pnt) == "UnitID")) {
    pnt <- pnt[, !(names(pnt) %in% "UnitID")]
  }

  res <- suppressMessages(st_join(sf::st_sf(pnt), vpu)$UnitID)

  if (all(is.na(res))) { # pnt is slightly outside of the vpu extent
    res <- vpu[which.min(sf::st_distance(vpu, pnt)), ]$UnitID
  }

  as.character(res)
}

#' find_state
#'
#' @param pnt an sf point object
#' @param abb logical return a state abbreviation?
#' @examples \dontrun{
#' pnt <- st_as_sf(data.frame(Lon = -107.2, Lat = 39.45),
#'   coords = c("Lon", "Lat"), crs = 4326)
#' }
find_state <- function(pnt, abb = FALSE) {
  state_data_sf <- sf::st_as_sf(map("state", plot = FALSE, fill = TRUE))
  res <- sf::st_transform(state_data_sf, sf::st_crs(pnt))
  res <- make_valid_geom_s2(res)

  res_intersects <- sf::st_intersects(res, pnt)

  res <- res[which(unlist(lapply(res_intersects, length)) > 0), ]
  state <- res$ID

  if (abb) {
    return(datasets::state.abb[tolower(datasets::state.name) == state])
  }
  return(state)
}

nhd_read_dbf <- function(state, dsn) {
  sf::st_read(normalizePath(gdb_path(state)), dsn)
}

# https://stackoverflow.com/a/9188972/3362993
long2UTM <- function(long) {
  (floor((long + 180) / 6) %% 60) + 1
}

#' Re-project to appropriate UTM zone
#'
#' @param sf_object an sf object
#'
#' @importFrom sf st_transform st_crs
#' @export
#' @return A transformed sf object
#'
#' @examples \dontrun{
#' data(gull)
#' gull_ <- gull$sp$NHDWaterbody
#' st_crs(gull_)
#' gull_ <- st_transform(gull_, 4326)
#' st_crs(gull_)
#' st_crs(toUTM(gull_[1, ]))
#' }
toUTM <- function(sf_object) {

  if (is.na(st_crs(sf_object)$epsg)) {
    sf_object <- st_transform(sf_object, crs = 4326)
  }

  if (sf::st_is_longlat(sf_object)) {
    utm_zone <- suppressWarnings(
      long2UTM(st_coordinates(st_centroid(st_union(sf_object)))[1])
    )
    crs      <- paste0("+proj=utm +zone=", utm_zone, " +datum=WGS84")

    sf::st_transform(sf_object, crs = crs)
  } else {
    sf_object
  }
}

compile_gpkg <- function(state) {
  # state = "DC"
  dest <- suppressWarnings(
    normalizePath(gsub(".gdb", ".gpkg", gdb_path(state)))
  )
  # file.exists(dest)
  sf::gdal_utils("vectortranslate",
    source = normalizePath(gdb_path(state)),
    destination = dest,
    options = c("-f", "GPKG")
  )
}

is_gpkg_installed <- function() {
  name <- rlang::quo("name")
  all(as.logical(dplyr::filter(sf::st_drivers(), name == "GPKG")[3:6]))
}

#' Convert a bounding box to polygon
#'
#' @param bbox object of class bbox from sf
#'
#' @importFrom sf st_as_sfc
#' @export
#' @return An sfc object from the sf package
#' @examples \dontrun{
#' library(sf)
#' wk <- wikilake::lake_wiki("Gull Lake (Michigan)")
#'
#' pnt <- st_as_sf(wk, coords = c("Lon", "Lat"), crs = 4326)
#' pnt <- st_transform(pnt, st_crs(vpu_shp))
#' qry <- nhd_plus_query(wk$Lon, wk$Lat,
#'   dsn = c("NHDWaterbody"), buffer_dist = 0.05)
#' wbd <- qry$sp$NHDWaterbody[which.max(st_area(qry$sp$NHDWaterbody)), ]
#' bbox2poly(st_bbox(wbd))
#' }
bbox2poly <- function(bbox) {
  sf::st_as_sfc(bbox)
}

# https://stackoverflow.com/a/14965990/3362993
has_7z <- function() {
  paths_7z <- c("7z",
    path.expand("~/usr/bin/7z"),
    "C:\\PROGRA~1\\7-Zip\\7za",
    "C:\\PROGRA~1\\7-Zip\\7z.exe")
  if (!any(nchar(Sys.which(paths_7z)) > 0)) {
    list(yes = FALSE, path = NA)
  } else {
    list(yes = TRUE, path = paths_7z[nchar(Sys.which(paths_7z)) > 0][1])
  }
}

get_utm_zone <- function(crs) {
  crs <- as.character(crs$proj4string)
  stringr::str_extract(crs, "(?!\\+zone=)(\\d+)(?=\\s\\+datum)")
}

stateabb2name <- function(abb) {
  # state <- "DC"
  key <- data.frame(sabb = c(datasets::state.abb, "DC", "AS", "GU", "MP", "PR", "VI"),
    sname =
      gsub(" ", "_",
        c(datasets::state.name, "District of Columbia", "American Samoa", "Guam", "Northern Mariana Islands", "Puerto Rico", "United States Virgin Islands")),
    stringsAsFactors = FALSE)

  res <- dplyr::filter(key, .data$sabb == abb)
  res$sname
}

albers_conic <- function() {
  # Albers Equal Area Conic
  "+proj=aea +lat_1=29.5 +lat_2=45.5 +lat_0=23 +lon_0=-96 +x_0=0 +y_0=0 +datum=NAD83 +units=m +no_defs"
}

#' Data and spatial polygons of the Great Lakes
#'
#' @importFrom purrr transpose
#' @importFrom sf st_as_sf
#' @export
#' @param spatial logical, return Great Lakes polygons?
#' @return A data frame of North America Great Lakes with
#'  optional geometry column
#' @examples
#' gl <- great_lakes()
#' \dontrun{
#' gl <- great_lakes(spatial = TRUE)
#' }
great_lakes <- function(spatial = FALSE) {
  res <- data.frame(
    GNIS_NAME = c("Lake Michigan", "Lake Erie", "Lake Huron", "Lake Ontario",
      "Lake Superior"),
    lon_dd = c(-87.0, -81.2, -82.4, -77.9, -87.5),
    lat_dd = c(44.0, 42.2, 44.8, 43.7, 47.0),
    stringsAsFactors = FALSE)

  if (spatial) {
    sp <- lapply(purrr::transpose(res), function(x) {
      res <- nhd_plus_query(
        x$lon_dd,
        x$lat_dd,
        dsn = "NHDWaterbody", buffer_dist = 0.3)$sp$NHDWaterbody

      if (nrow(res) > 0) {
        res <- st_transform(res, albers_conic())
      }
    })
    sp  <- do.call(what = rbind, args = sp)
    res <- st_as_sf(dplyr::left_join(res, sp, by = "GNIS_NAME"))
  }

  res
}

#' @importFrom utils capture.output
st_read_custom <- function(x, pretty = FALSE, wkt_filter = NA, ...) {
  if (isTRUE(pretty)) {
    msg <- capture.output(res <- sf::st_read(x, ...))
    if (length(msg) > 0) {
      msg <-
        stringr::str_extract(msg[1],
          "([?:`].*)(?=' from data source)")
      msg <- substring(msg, 2, nchar(msg))
      message(paste0("Reading layer '", msg, "'"))
      res
    } else {
      res
    }
  } else {
    if (!is.na(wkt_filter)) {
      sf::st_read(x, wkt_filter = wkt_filter, ...)
    } else {
      sf::st_read(x, ...)
      # do.call(sf::st_read, c("dsn" = x, arguments))
    }
  }
}

align_names <- function(to, from) {
  res <- rep(NA, length(to))
  for (i in seq_along(to)) {
    if (tolower(to[i]) %in% tolower(from)) {
      res[i] <- from[which(tolower(to[i]) == tolower(from))]
    } else {
      res[i] <- to[i]
    }
  }
  res
}

# https://github.com/r-spatial/s2/issues/99#issuecomment-827776431
make_valid_geom_s2 <- function(sf_object) {
  sf_s2 <- s2::s2_rebuild(
    s2::as_s2_geography(sf_object, check = FALSE),
    options = s2::s2_options(
      edge_type = "undirected", split_crossing_edges = TRUE, validate = TRUE
    )
  )
  # all(s2::s2_is_valid(vpu_s2))

  sf::st_geometry(sf_object) <- sf::st_as_sfc(sf_s2)
  sf_object
}

st_line_sample_4326 <- function(lines) {
  sf::st_transform(st_cast(
    sf::st_line_sample(sf::st_transform(lines, 3395), sample = 0),
    "POINT"), 4326)
}
